packages/rocketchat-otr/client/rocketchat.otr.room.js

```diff
@@ -90,155 +73,115 @@ RocketChat.OTR.Room = class {
-				name: "RSA-OAEP",
-			},
-			publicKey,
-			new TextEncoder("UTF-8").encode(message) //ArrayBuffer of data you want to encrypt
-		)
-		.then((encrypted) => {
-			return this.bytesToHexString(encrypted);
+	generateKeyPair(refreshKeys) {
+		// Generate an ephemeral key pair.
+		return window.crypto.subtle.generateKey({
+			name: 'ECDH',
+			namedCurve: 'P-256'
+		}, false, ['deriveKey', 'deriveBits']).then((keyPair) => {
+			this.keyPair = keyPair;
+			return crypto.subtle.exportKey('spki', keyPair.publicKey);
+		})
+		.then((exportedPublicKey) => {
+			this.exportedPublicKey = exportedPublicKey;
 		})
 		.catch((err) => {
-			console.log(err);
-			return message;
+			console.error(err);
 		});
 	}
 
-	decryptRSA(message) {
-		return window.crypto.subtle.decrypt({
-				name: "RSA-OAEP",
-			},
-			this.privateKey,
-			new this.hexStringToUint8Array(message) //ArrayBuffer of the data
-		)
-		.then((decrypted) => {
-			//returns an ArrayBuffer containing the decrypted data
-			return new TextDecoder("UTF-8").decode(new Uint8Array(decrypted));
-		})
-		.catch((err) => {
-			console.log(err);
-			return message;
+	importPublicKey(publicKey) {
+		return window.crypto.subtle.importKey('spki', this.hexStringToUint8Array(publicKey), {
+			name: 'ECDH',
+			namedCurve: 'P-256'
+		}, false, []).then((peerPublicKey) => {
+			return crypto.subtle.deriveBits({
+				name: 'ECDH',
+				namedCurve: 'P-256',
+				public: peerPublicKey
+			}, this.keyPair.privateKey, 256);
+		}).then((bits) => {
+			return crypto.subtle.digest({
+				name: 'SHA-256'
+			}, bits);
+		}).then((hashedBits) => {
+			// We truncate the hash to 128 bits.
+			var sessionKeyData = new Uint8Array(hashedBits).slice(0, 16);
+			return crypto.subtle.importKey('raw', sessionKeyData, {
+				name: 'AES-GCM'
+			}, false, ['encrypt', 'decrypt']);
+		}).then((sessionKey) => {
+			// Session key available.
+			this.sessionKey = sessionKey;
 		});
 	}
 
-	getSharedSecret() {
-		return window.crypto.subtle.generateKey({
-				name: "AES-CBC",
-				length: 256, //can be  128, 192, or 256
-			},
-			true, //whether the key is extractable (i.e. can be used in exportKey)
-			["encrypt", "decrypt"] //can be "encrypt", "decrypt", "wrapKey", or "unwrapKey"
-		)
-		.then((sharedSecret) => {
-			this.sharedSecret = sharedSecret;
-			// export public key
-			return window.crypto.subtle.exportKey(
-				"jwk", //can be "jwk" (public or private), "spki" (public only), or "pkcs8" (private only)
-				sharedSecret //can be a publicKey or privateKey, as long as extractable was true
-			)
-			.then((sharedSecretJWK) => {
-				this.sharedSecretJWK = sharedSecretJWK;
-			})
-		})
-	}
+	encrypt(message) {
+		// var clearText = new Uint8Array(message);
+		var clearText = new TextEncoder("UTF-8").encode(message);
+		var userId = new TextEncoder("UTF-8").encode(this.userId);
+		this.serial++;
+		var data = new Uint8Array(1 + 1 + userId.length + clearText.length);
+		data[0] = this.serial;
```

Comment:
Serial has to be an array.

You can use something like the following to have serial inside an array:

```javascript
function enlargeArray(array, newLength) {
    var newArray = new Uint8Array(newLength);
    newArray.set(array, newLength - array.length);
    return newArray;
  }

  function increment(array) {
    for (var i = array.length - 1; i >= 0; i--) {
      array[i]++;
      // Overflow happened, we have to carry over.
      if (array[i] === 0) {
        continue;
      }
      else {
        // We are done.
        return array;
      }
    }

    // We have to enlarge the array and carry over.
    array = enlargeArray(array, array.length + 1);
    array[0] = 1;
    return array;
  }
```